zerver/tests/fixtures/markdown_test_cases.json

```diff
@@ -853,6 +853,12 @@
       "expected_output": "<ol>\n<li>\n<p>A</p>\n</li>\n<li>\n<p>B</p>\n</li>\n<li>\n<p>C</p>\n</li>\n<li>\n<p>D<br>\nordinary paragraph</p>\n</li>\n<li>\n<p>AA</p>\n</li>\n<li>\n<p>BB</p>\n</li>\n</ol>",
       "marked_expected_output": "<ol>\n<li><p>A</p>\n</li>\n<li><p>B</p>\n</li>\n<li><p>C</p>\n</li>\n<li><p>D<br>\nordinary paragraph</p>\n</li>\n<li><p>AA</p>\n</li>\n<li><p>BB</p>\n</li>\n</ol>",
       "text_content": "1. A\n2. B\n3. C\n4. D\nordinary paragraph\n5. AA\n6. BB"
+    },
+    {
+      "name": "zpoilers_fenced_spoiler",
+      "input": "```spoiler header\ncontent\n```\noutside spoiler\n",
+      "expected_output": "<div class=\"zpoiler-block\"><div class=\"zpoiler-header\"><a class=\"zpoiler-button\" aria-expanded=\"false\"><span class=\"zpoiler-arrow zpoiler-button-closed\"></span></a>\n\n<p>header</p>\n</div><div class=\"zpoiler-content zpoiler-content-closed\" aria-hidden=\"true\">\n\n<p>content</p>\n</div></div>\n\n<p>outside spoiler</p>",
+      "marked_expected_output": "<div class=\"zpoiler-block\"><div class=\"zpoiler-header\"><a class=\"zpoiler-button\" aria-expanded=\"false\"><span class=\"zpoiler-arrow zpoiler-button-closed\"></span></a>\n\nheader\n\n</div><div class=\"zpoiler-content zpoiler-content-closed\" aria-hidden=\"true\">\n\ncontent\n\n</div></div>\n\n<p>outside spoiler</p>"
```

Comment:
Can you add a few more tests:
* A spoiler inside a block quote.
* A spoiler where no heading was provided.
* A spoiler with markdown syntax in the header (E.g. an emoji and a link).

Reply:
I just checked on this. Links to images that end in one of the extensions in the [`is_image` function in bugdown](https://github.com/zulip/zulip/blob/master/zerver/lib/bugdown/__init__.py#L671) still work fine, because they're expanded into an inline image. This *is* an issue for other pages which unfurl content (in a format similar to open graph cards) and append it to the bottom. Zulip appears to have [specific logic to unroll Twitter and Youtube links](https://github.com/zulip/zulip/blob/master/static/js/markdown.js#L28); as far as I can tell, those are the only two sites that it does this for, but those links are definitely affected and appear outside the spoiler block.

We'd be fine just testing the known problem for the time being. What are your ideas for fixing this in the future?

Reply:
My best idea for fixing it is to have the logic that adds the previews look at whether the link was within a `spoiler_block` via the tree structure, and if so, append to end of that rather than the end of the message.